8Knot/cache_manager/cache_facade.py

```diff
@@ -0,0 +1,236 @@
+
+
+def get_uncached(func_name: str, repolist: list[int]) -> list[int]:  # or None
+    """
+    Checks bookkeeping data to find, for a given querying function, which
+    repos' data are noted as present in cache vs. those that arent.
+
+    Returns a list of repos that AREN'T resident in cache.
+    """
+    with pg.connect(env_db_connection_string) as cache_conn:
+        with cache_conn.cursor() as cache_cur:
+            composed_query = pg_sql.SQL(
+                """
+                SELECT cb.repo_id
+                FROM cache_bookkeeping cb
+                WHERE cb.cache_func = '{cache_func_name}' AND cb.repo_id in %s
+                """.format(
+                    cache_func_name=func_name
+                )
+            ).as_string(cache_conn)
+
+            # exec query
+            cache_cur.execute(query=composed_query, vars=(tuple(repolist),))
+
+            # get list of cached repos
+            already_cached: list[tuple] = cache_cur.fetchall()
+
+            # process list of single-value tuples to get list of values.
+            # looks like: [(val1,), (val2,), ...]
+            already_cached: set[int] = set([v[0] for v in already_cached])
+
+            # repos that are already cached will be removed from repolist set,
+            # leaving uncached remaining.
+            not_cached: list[int] = list(set(repolist) - already_cached)
+
+            return not_cached
+
+
+def caching_wrapper(query_name: str, func_name: str, table_name: str, query: str, repolist: list[int]) -> None:
+    """Combines steps of (1) identifying which repos aren't already cached and
+    (2) querying + caching repos those repos.
+
+    Args:
+        query_name (str): name of querying function's file
+        func_name (str): literal name of querying function for bookkeeping
+        table_name (str): which table to insert results into
+        query (str): sql query as a string
+        repolist (list[int]): list of repos requested by user.
+
+    Raises:
+        Exception: If a step fails, will print exception and re-raise.
+
+    Returns:
+        _type_: None
+    """
+    try:
+        # STEP 1: Which repos need to be queried for?
+        #           some might already be in cache.
+        uncached_repos: list[int] | None = get_uncached(func_name=func_name, repolist=repolist)
+        if not uncached_repos:
+            logging.warning(f"{query_name}_DATA_QUERY - ALL REQUESTED REPOS IN CACHE")
+            return 0
+        else:
+            logging.warning(f"{query_name}_DATA_QUERY - CACHING {len(uncached_repos)} NEW REPOS")
+            uncached_repos: tuple[tuple] = tuple([tuple(uncached_repos)])
+
+        # STEP 2: Query for those repos
+        logging.warning(f"{query_name}_DATA_QUERY - EXECUTING CACHING QUERY")
+        cache_query_results(
+            # TODO: should get this from the environment
+            db_connection_string="dbname=padres user=cali password=!baseball21 host=chaoss.tv port=5432",
```

Comment:
yeah lets change this asap I dont want to have any passwords or anything in git logs

Reply:
oop